File: 04.Simulation模拟/leetcode_1807_evaluate.py
from typing import List

class Solution:
    def evaluate(self, s: str, knowledge: List[List[str]]) -> str:
        mp = dict()
        for k,v in knowledge:
            mp[k] = v
        
        res = ""
        cur = ""
        for i in s:
            if i == ')':
                res += mp.get(cur[1:], '?')
                cur = ""
            elif i == '(' or len(cur) > 0:
                cur += i
            else:
                res += i
        return res

if __name__ == "__main__":
    sl = Solution()
    s = "(name)is(age)yearsold"
    knowledge = [["name","bob"],["age","two"]]
    print(sl.evaluate(s, knowledge))

File: 04.Simulation模拟/leetcode_1807_evaluate.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_map>
using namespace std;

class Solution {
public:
    string evaluate(string s, vector<vector<string>>& knowledge) {
        unordered_map<string, string> mp;
        for (auto& k : knowledge){
            mp[k[0]] = k[1];
        }

        string res = "";
        string cur = "";
        for (char c : s){
            if (c == ')'){
                res += (mp.count(cur.substr(1)) ? mp[cur.substr(1)] : "?");
                cur = "";
            }
            else if (c == '(' || cur.size() > 0){
                cur += c;
            }
            else{
                res += c;
            }
        }
        return res;
    }
};

int main(){
    Solution sl;
    string s = "(name)is(age)yearsold";
    vector<vector<string>> knowledge = {{"name", "bob"}, {"age", "two"}};
    cout << sl.evaluate(s, knowledge) << endl;
}
